src/bigint.rs

```diff
@@ -11,14 +11,24 @@ pub fn sample(bit_size: usize) -> BigInt {
     BigInt::from(&*buf) >> (bytes * 8 - bit_size)
 }
 
+/// Sample a number of bit length bit_size
 pub fn sample_with_rng(rng: &mut (impl CryptoRng + RngCore), bit_size: usize) -> BigInt {
     if bit_size == 0 {
         return BigInt::zero();
     }
+
     let bytes = (bit_size - 1) / 8 + 1;
     let mut buf: Vec<u8> = vec![0; bytes];
+
     rng.fill_bytes(&mut buf);
-    BigInt::from(&*buf) >> (bytes * 8 - bit_size)
+
+    let mut p = BigInt::from(buf.as_ref());
+    p >>= bytes * 8 - bit_size;
+
+    // Set the MSB to 1 to get full bit length
+    p.setbit(bit_size - 1);
```

Comment:
My first instinct is to object to this on the grounds that it eliminates half of all possible numbers but I guess we need prime candidates to have a minimum size for security.

Reply:
We lose one bit of security but both safe primes get to be of the safe size. I noticed that some implementations also did this. The code this was adapted from did rejection sampling instead which was much slower. But now that you have mentioned this, even if the two safe primes are unbalanced, I'm not sure if we need this to prevent anything.

Reply:
(For future reference, the probability of them being very unbalanced drops exponentially)